lib/twilio-ruby/jwt/jwt.rb

```diff
@@ -0,0 +1,55 @@
+module Twilio
+  module JWT
+    class BaseJWT
+      def initialize(secret_key: nil, issuer: nil, subject: nil, algorithm: 'HS256', nbf: nil, ttl: 3600, valid_until: nil)
+        @secret_key = secret_key
+        @issuer = issuer
+        @subject = subject
+        @algorithm = algorithm
+        @nbf = nbf
+        @ttl = ttl
+        @valid_until = valid_until
+      end
+
+      def _generate_headers
+        return {}
+      end
+
+      def _generate_payload
+        fail NotImplementedError
+      end
+
+      def headers
+        headers = self._generate_headers.clone()
+        headers[:typ] = 'JWT'
+        headers[:alg] = @algorithm
+        return headers
+      end
+
+      def payload
+        payload = self._generate_payload.clone()
+
+        payload[:iss] = @issuer
+        payload[:nbf] = @nbf || Time.now.to_i
+
+        if @valid_until.nil?
+          payload[:exp] = Time.now.to_i + @ttl
+        else
+          payload[:exp] = @valid_until
+        end
+
+        payload[:sub] = @subject unless @subject.nil?
+
+        return payload
+      end
+
+      def to_jwt
+        if @secret_key.nil?
+          raise ArgumentError, 'JWT does not have a signing key'
```

Comment:
https://robots.thoughtbot.com/ruby-2-keyword-arguments#required-keyword-arguments

We would need to drop Ruby 2 (it's only supported in 2.1) if we want to keep it as a keyword argument.